server/src/main/java/org/elasticsearch/search/rank/context/RankFeaturePhaseRankCoordinatorContext.java

```diff
@@ -74,16 +74,12 @@ public void computeRankScoresForGlobalResults(
         RankFeatureDoc[] featureDocs = extractFeatureDocs(rankSearchResults);
 
         // generate the final `topResults` results, and pass them to fetch phase through the `rankListener`
-        if (featureDocs.length == 0) {
-            rankListener.onResponse(new RankFeatureDoc[0]);
-        } else {
-            computeScores(featureDocs, rankListener.delegateFailureAndWrap((listener, scores) -> {
-                for (int i = 0; i < featureDocs.length; i++) {
-                    featureDocs[i].score = scores[i];
-                }
-                listener.onResponse(featureDocs);
-            }));
-        }
+        computeScores(featureDocs, rankListener.delegateFailureAndWrap((listener, scores) -> {
+            for (int i = 0; i < featureDocs.length; i++) {
+                featureDocs[i].score = scores[i];
+            }
+            listener.onResponse(featureDocs);
+        }));
```

Comment:
I wonder if we can short-circuit the inference call on zero docs in the ML code, after the inference ID has been validated but before we make the call to the external inference service?

Reply:
In this particular case, I can add a short-circuit right in the `TextSimilarityRankFeaturePhaseRankCoordinatorContext` - which is a good idea. 

The bigger concern that I'm ambivalent on is removing this short circuit in the parent class, so every implementation of this context has to short circuit going forward.

Reply:
I think this bug shows us that the short-circuit logic cannot be generalized to the level that we can perform it with a catch-all implementation in the parent class. In some cases, like in `RandomRankFeaturePhaseRankCoordinatorContext`, the catch-all implementation in the parent class works fine. But in other cases, like in rerankers that call an inference endpoint, it doesn't work because the inference endpoint is not validated. That leads me to the conclusion that we should move the short-circuit logic to the child classes so that it can be customized as necessary for the situation.

Reply:
Just double checking if this handling logic still accounts for the use case where the rank feature extraction phase returns an empty array.

While @maxhniebergall 's [comment](https://github.com/elastic/elasticsearch/issues/111934#issuecomment-2291690065) is on point (we should fail fast if the field doesn't exist in the mapping), it _is_ a valid case that the first-stage retrieval returns documents that don't have the specified field.

Reply:
@demjened What is the expected behavior if none of the documents returned by the first-stage retrieval contain the field to be reranked on? Does reranking become a no-op in that case?

Reply:
++ @demjened it would be helpful to elaborate a bit more on this requirement, thanks!

Reply:
We've [discussed](https://elastic.slack.com/archives/C06MTP45DL6/p1716365692136049?thread_ts=1716321808.304609&cid=C06MTP45DL6) some alternatives, AFAIK the current expectation is
- Filter documents that don't have `field` before reranking (this is implemented)
- Document this behavior (this is still missing from [this page](https://www.elastic.co/guide/en/elasticsearch/reference/current/retriever.html#text-similarity-reranker-retriever))

Reply:
Thanks for the catch. You are correct - I added a test to validate reranking behavior on a field that exists but is not returned. 

Note: by short circuiting here, we will still return 0 results on a nonexistent field. My suggestion is that we keep that behavior for this PR and decide whether we want to open a followup defect for this. It would require field lookups and we may want to discuss whether we want to add that to this context, but either way it's a larger change.

Reply:
Agreed, we should iterate on nonexistent field handling in a separate PR. In that case, we may want to use the [index.query.parse.allow_unmapped_fields index setting](https://github.com/elastic/elasticsearch/blob/6dbf8d59e578c9cd4cea77b3f326aeaf9c49758b/server/src/main/java/org/elasticsearch/index/IndexSettings.java#L81) to determine if we should throw on a nonexistent field.